// @inworld/framework-nodejs/build/primitives/safety/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicName = void 0;
/**
 * @enum TopicName
 * Enumeration of content topics that can be detected and filtered.
 */
var TopicName;
(function (TopicName) {
    /** Politics and political discussions */
    TopicName["Politics"] = "Politics";
    /** Content related to substance use and drugs */
    TopicName["SubstanceUse"] = "SubstanceUse";
    /** Religious content and discussions */
    TopicName["Religion"] = "Religion";
    /** Content related to self-harm, suicide, or self-injury */
    TopicName["UnconditionalSelfHarm"] = "UnconditionalSelfHarm";
    /** Inappropriate content involving minors */
    TopicName["UnconditionalSexualMinors"] = "UnconditionalSexualMinors";
    /** Content promoting hate groups or extremist organizations */
    TopicName["UnconditionalHateGroup"] = "UnconditionalHateGroup";
    /** Adult sexual content */
    TopicName["AdultSexual"] = "AdultSexual";
    /** Violent content, descriptions of harm, or threatening language */
    TopicName["Violence"] = "Violence";
    /** Content related to alcohol consumption */
    TopicName["Alcohol"] = "Alcohol";
    /** Profane language, swear words, and obscenities */
    TopicName["Profanity"] = "Profanity";
})(TopicName || (exports.TopicName = TopicName = {}));
