// @inworld/framework-nodejs/build/primitives/intent/intent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Intent = void 0;
const common_1 = require("../../common");
const expose_binary_1 = require("../../expose_binary");
/**
 * Class representing an intent.
 * Intents define recognizable user requests or commands with example phrases.
 */
class Intent {
    /**
     * Creates a new Intent instance.
     *
     * @param props - Intent properties including name and example phrases
     */
    constructor(props) {
        this.intent = expose_binary_1.IntentFunctions.new();
        expose_binary_1.IntentFunctions.setName(this.intent, props.name);
        this.vectorPhrases = new common_1.VectorString(props.phrases);
        expose_binary_1.IntentFunctions.setPhrases(this.intent, this.vectorPhrases.getExternal());
    }
    /**
     * Gets the name of this intent.
     *
     * @returns Name of the intent
     */
    getName() {
        return expose_binary_1.IntentFunctions.getName(this.intent);
    }
    /**
     * Gets the example phrases for this intent.
     *
     * @returns Array of example phrases
     */
    getPhrases() {
        const vectorPhrases = expose_binary_1.IntentFunctions.getPhrases(this.intent);
        const phrases = common_1.VectorString.toArray(vectorPhrases);
        return phrases;
    }
    /**
     * Returns the external reference to the native intent implementation.
     *
     * @returns External reference object
     * @internal
     */
    getExternal() {
        return this.intent;
    }
    /**
     * Cleans up resources associated with this intent.
     */
    destroy() {
        if (this.vectorPhrases) {
            this.vectorPhrases.destroy();
            this.vectorPhrases = null;
        }
        if (this.intent) {
            expose_binary_1.IntentFunctions.delete(this.intent);
            this.intent = null;
        }
    }
}
exports.Intent = Intent;
