// @inworld/framework-nodejs/build/primitives/intent/intent_match.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntentMatch = void 0;
const expose_binary_1 = require("../../expose_binary");
/**
 * Class representing a matched intent.
 * Contains information about an intent that was detected in user input.
 */
class IntentMatch {
    /**
     * Creates a new IntentMatch instance.
     *
     * @param intentMatch - External reference to the native intent match implementation
     * @internal
     */
    constructor(intentMatch) {
        this.external = null;
        this.external = intentMatch;
    }
    /**
     * Gets the name of the matched intent.
     *
     * @returns Name of the matched intent
     */
    getName() {
        return expose_binary_1.IntentMatchFunctions.getIntentName(this.external);
    }
    /**
     * Gets the confidence score of the intent match.
     *
     * @returns Confidence score (typically between 0 and 1)
     */
    getScore() {
        return expose_binary_1.IntentMatchFunctions.getScore(this.external);
    }
    /**
     * Returns the external reference to the native intent match implementation.
     *
     * @returns External reference object
     * @internal
     */
    getExternal() {
        return this.external;
    }
    /**
     * Cleans up resources associated with this intent match.
     */
    destroy() {
        if (this.external) {
            expose_binary_1.IntentMatchFunctions.delete(this.external);
            this.external = null;
        }
    }
}
exports.IntentMatch = IntentMatch;
