// @inworld/framework-nodejs/build/common/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomOutputDataType = exports.CustomInputDataType = void 0;
/**
 * Enum representing the types of data that can be processed in the graph.
 */
var CustomInputDataType;
(function (CustomInputDataType) {
    CustomInputDataType["TEXT"] = "TEXT";
    CustomInputDataType["JSON"] = "JSON";
    CustomInputDataType["TEXT_STREAM"] = "TEXT_STREAM";
    CustomInputDataType["TTS_STREAM"] = "TTS_STREAM";
    CustomInputDataType["KNOWLEDGE_RECORDS"] = "KNOWLEDGE_RECORDS";
})(CustomInputDataType || (exports.CustomInputDataType = CustomInputDataType = {}));
/**
 * Enum representing the types of data that can be output from the graph.
 */
var CustomOutputDataType;
(function (CustomOutputDataType) {
    CustomOutputDataType["TEXT"] = "TEXT";
    CustomOutputDataType["JSON"] = "JSON";
    CustomOutputDataType["AUDIO"] = "AUDIO";
    CustomOutputDataType["CHAT_MESSAGES"] = "CHAT_MESSAGES";
})(CustomOutputDataType || (exports.CustomOutputDataType = CustomOutputDataType = {}));
