// @inworld/framework-nodejs/build/graph/nodes/builtin/node_execution_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMNodeExecutionConfig = exports.IntentNodeExecutionConfig = exports.TTSNodeExecutionConfig = exports.STTNodeExecutionConfig = exports.ExecutionConfig = void 0;
const voice_1 = require("../../../common/voice/voice");
const expose_binary_1 = require("../../../expose_binary");
const intent_1 = require("../../../primitives/intent");
const text_generator_1 = require("../../../primitives/llm/text_generator");
const config_1 = require("../../../primitives/stt/config");
const config_2 = require("../../../primitives/tts/config");
/**
 * Represents the execution config.
 * @internal
 */
class ExecutionConfig {
    /**
     * Constructs a new execution config.
     *
     * @param external - The external reference.
     * @param props - The properties.
     */
    constructor(external, props) {
        var _a;
        this.external = external;
        this.reportToClient = (_a = props === null || props === void 0 ? void 0 : props.reportToClient) !== null && _a !== void 0 ? _a : false;
    }
    /**
     * Gets the external reference.
     *
     * @returns The external reference.
     */
    getExternal() {
        return this.external;
    }
    /**
     * Destroys the execution config.
     */
    destroy() {
        this.external = null;
    }
}
exports.ExecutionConfig = ExecutionConfig;
/**
 * Represents the STT node execution config.
 * @internal
 */
class STTNodeExecutionConfig extends ExecutionConfig {
    /**
     * Constructs a new STT node execution config.
     *
     * @param props - The properties.
     */
    constructor(props) {
        super(expose_binary_1.ExecutionConfigFunctions.newSTTNodeExecutionConfig(), props);
        try {
            this.speechRecognitionConfig = new config_1.SpeechRecognitionConfig(props === null || props === void 0 ? void 0 : props.speechRecognitionConfig);
            expose_binary_1.ExecutionConfigFunctions.setSpeechRecognitionConfigForSTTNodeExecutionConfig(this.external, this.speechRecognitionConfig.getExternal());
            expose_binary_1.ExecutionConfigFunctions.setReportToClientForSTTNodeExecutionConfig(this.external, this.reportToClient);
        }
        catch (error) {
            this.destroy();
            throw error;
        }
    }
    /**
     * Destroys the STT node execution config.
     */
    destroy() {
        var _a;
        (_a = this.speechRecognitionConfig) === null || _a === void 0 ? void 0 : _a.destroy();
        this.speechRecognitionConfig = null;
        if (this.external) {
            expose_binary_1.ExecutionConfigFunctions.deleteSTTNodeExecutionConfig(this.external);
        }
        super.destroy();
    }
}
exports.STTNodeExecutionConfig = STTNodeExecutionConfig;
/**
 * Represents the TTS node execution config.
 * @internal
 */
class TTSNodeExecutionConfig extends ExecutionConfig {
    constructor(props) {
        super(expose_binary_1.ExecutionConfigFunctions.newTTSNodeExecutionConfig(), props);
        let voice;
        let speechSynthesisConfig;
        try {
            voice = new voice_1.Voice(props.speakerId);
            speechSynthesisConfig = new config_2.SpeechSynthesisConfig(props.synthesisConfig);
            expose_binary_1.ExecutionConfigFunctions.setVoiceForTTSNodeExecutionConfig(this.external, voice.getExternal());
            expose_binary_1.ExecutionConfigFunctions.setSpeechSynthesisConfigForTTSNodeExecutionConfig(this.external, speechSynthesisConfig.getExternal());
            expose_binary_1.ExecutionConfigFunctions.setReportToClientForTTSNodeExecutionConfig(this.external, this.reportToClient);
            this.voice = voice;
            this.speechSynthesisConfig = speechSynthesisConfig;
        }
        catch (error) {
            this.destroy();
            throw error;
        }
    }
    /**
     * Destroys the TTS node execution config.
     */
    destroy() {
        var _a, _b;
        (_a = this.speechSynthesisConfig) === null || _a === void 0 ? void 0 : _a.destroy();
        this.speechSynthesisConfig = null;
        (_b = this.voice) === null || _b === void 0 ? void 0 : _b.destroy();
        this.voice = null;
        if (this.external) {
            expose_binary_1.ExecutionConfigFunctions.deleteTTSNodeExecutionConfig(this.external);
        }
        super.destroy();
    }
}
exports.TTSNodeExecutionConfig = TTSNodeExecutionConfig;
/**
 * Represents the intent node execution config.
 * @internal
 */
class IntentNodeExecutionConfig extends ExecutionConfig {
    /**
     * Constructs a new intent node execution config.
     *
     * @param props - The properties.
     */
    constructor(props) {
        super(expose_binary_1.ExecutionConfigFunctions.newIntentNodeExecutionConfig(), props);
        try {
            this.matcherConfig = new intent_1.IntentMatcherCreationConfig(props.matcherConfig);
            expose_binary_1.ExecutionConfigFunctions.setMatcherConfigForIntentNodeExecutionConfig(this.external, this.matcherConfig.getExternal());
            expose_binary_1.ExecutionConfigFunctions.setReportToClientForIntentNodeExecutionConfig(this.external, this.reportToClient);
        }
        catch (error) {
            this.destroy();
            throw error;
        }
    }
    /**
     * Destroys the intent node execution config.
     */
    destroy() {
        var _a;
        (_a = this.matcherConfig) === null || _a === void 0 ? void 0 : _a.destroy();
        this.matcherConfig = null;
        if (this.external) {
            expose_binary_1.ExecutionConfigFunctions.deleteIntentNodeExecutionConfig(this.external);
        }
        super.destroy();
    }
}
exports.IntentNodeExecutionConfig = IntentNodeExecutionConfig;
/**
 * Represents the LLM node execution config.
 * @internal
 */
class LLMNodeExecutionConfig extends ExecutionConfig {
    /**
     * Constructs a new LLM node execution config.
     *
     * @param props - The properties.
     */
    constructor(props) {
        super(expose_binary_1.ExecutionConfigFunctions.newLLMNodeExecutionConfig(), props);
        try {
            this.textGenerationConfig = new text_generator_1.TextGenerationConfig(props.textGenerationConfig);
            expose_binary_1.ExecutionConfigFunctions.setTextGenerationConfigForLLMNodeExecutionConfig(this.external, this.textGenerationConfig.getExternal());
            expose_binary_1.ExecutionConfigFunctions.setReportToClientForLLMNodeExecutionConfig(this.external, this.reportToClient);
        }
        catch (error) {
            this.destroy();
            throw error;
        }
    }
    /**
     * Destroys the LLM node execution config.
     */
    destroy() {
        var _a;
        (_a = this.textGenerationConfig) === null || _a === void 0 ? void 0 : _a.destroy();
        this.textGenerationConfig = null;
        if (this.external) {
            expose_binary_1.ExecutionConfigFunctions.deleteLLMNodeExecutionConfig(this.external);
        }
        super.destroy();
    }
}
exports.LLMNodeExecutionConfig = LLMNodeExecutionConfig;
