// @inworld/framework-nodejs/build/graph/nodes/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Node = void 0;
const common_1 = require("../../common");
const data_types_1 = require("../data_types");
/**
 * Base implementation of a graph node.
 * Provides common functionality for all node types.
 * @internal
 */
class Node {
    /**
     * Creates a new Node instance.
     *
     * @param id - Unique identifier for the node
     * @param external - Optional external reference to a native node implementation
     */
    constructor(id, external) {
        /** External reference to the node's native implementation */
        this.external = null;
        this.id = id;
        this.external = external;
    }
    /**
     * Returns the external reference to the node's native implementation.
     *
     * @returns External reference object
     * @internal
     */
    getExternal() {
        return this.external;
    }
    /**
     * Checks the status of data and throws an error if there's a problem.
     *
     * @param data - External reference to the data to check
     * @throws InworldError if the data contains an error status
     */
    checkStatus(data) {
        const status = data_types_1.BaseData.getError(data);
        if (status) {
            throw new common_1.InworldError(`${this.constructor.name} error`, {
                status,
            });
        }
    }
    /**
     * Cleans up resources associated with this node.
     */
    destroy() {
        this.external = null;
    }
}
exports.Node = Node;
