// @inworld/framework-nodejs/build/primitives/knowledge/local_knowledge_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalKnowledgeConfig = void 0;
const expose_binary_1 = require("../../expose_binary");
const knowledge_compile_config_1 = require("./knowledge_compile_config");
const knowledge_get_config_1 = require("./knowledge_get_config");
/**
 * Class for local knowledge system configuration
 * Configuration for local knowledge system
 * @internal
 */
class LocalKnowledgeConfig {
    /**
     * Creates a new LocalKnowledgeConfig instance
     * @param config Configuration parameters
     */
    constructor(config) {
        this.external = expose_binary_1.LocalKnowledgeConfigFunctions.new();
        // Create and set knowledge get config
        this.knowledgeGetConfig = new knowledge_get_config_1.KnowledgeGetConfig(config.knowledgeGetConfig);
        expose_binary_1.LocalKnowledgeConfigFunctions.setKnowledgeGetConfig(this.external, this.knowledgeGetConfig.getExternal());
        // Create and set knowledge compile config
        this.knowledgeCompileConfig = new knowledge_compile_config_1.KnowledgeCompileConfig(config.knowledgeCompileConfig);
        expose_binary_1.LocalKnowledgeConfigFunctions.setKnowledgeCompileConfig(this.external, this.knowledgeCompileConfig.getExternal());
    }
    /**
     * Gets the reference to the native object
     * @returns Native object reference
     */
    getExternal() {
        return this.external;
    }
    /**
     * Destroys the native object
     */
    destroy() {
        if (this.knowledgeGetConfig) {
            this.knowledgeGetConfig.destroy();
        }
        if (this.knowledgeCompileConfig) {
            this.knowledgeCompileConfig.destroy();
        }
        if (this.external) {
            expose_binary_1.LocalKnowledgeConfigFunctions.delete(this.external);
            this.external = null;
        }
    }
}
exports.LocalKnowledgeConfig = LocalKnowledgeConfig;
