// @inworld/framework-nodejs/build/primitives/knowledge/remote_knowledge_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteKnowledgeConfig = void 0;
const expose_binary_1 = require("../../expose_binary");
const knowledge_compile_config_1 = require("./knowledge_compile_config");
const knowledge_get_config_1 = require("./knowledge_get_config");
/**
 * Class for remote knowledge system configuration
 * @internal
 */
class RemoteKnowledgeConfig {
    /**
     * Creates a new RemoteKnowledgeConfig instance
     * @param config Configuration parameters
     */
    constructor(config) {
        this.knowledgeGetConfig = null;
        this.external = expose_binary_1.RemoteKnowledgeConfigFunctions.new();
        // Set API key
        expose_binary_1.RemoteKnowledgeConfigFunctions.setApiKey(this.external, config.apiKey);
        // Create and set knowledge compile config
        this.knowledgeCompileConfig = new knowledge_compile_config_1.KnowledgeCompileConfig(config.knowledgeCompileConfig);
        expose_binary_1.RemoteKnowledgeConfigFunctions.setKnowledgeCompileConfig(this.external, this.knowledgeCompileConfig.getExternal());
        // Create and set knowledge get config if provided
        if (config.knowledgeGetConfig) {
            this.knowledgeGetConfig = new knowledge_get_config_1.KnowledgeGetConfig(config.knowledgeGetConfig);
            // Note: The C++ bindings might not support this yet
            // Use type assertion to avoid TypeScript errors
            const configFunctions = expose_binary_1.RemoteKnowledgeConfigFunctions;
            if (typeof configFunctions.setKnowledgeGetConfig === 'function') {
                configFunctions.setKnowledgeGetConfig(this.external, this.knowledgeGetConfig.getExternal());
            }
        }
    }
    /**
     * Gets the reference to the native object
     * @returns Native object reference
     */
    getExternal() {
        return this.external;
    }
    /**
     * Destroys the native object
     */
    destroy() {
        if (this.knowledgeCompileConfig) {
            this.knowledgeCompileConfig.destroy();
        }
        if (this.knowledgeGetConfig) {
            this.knowledgeGetConfig.destroy();
        }
        if (this.external) {
            expose_binary_1.RemoteKnowledgeConfigFunctions.delete(this.external);
            this.external = null;
        }
    }
}
exports.RemoteKnowledgeConfig = RemoteKnowledgeConfig;
